Use built-in alertmanager config if no fallback config is provided

JohnnyQQQQ opened the pull request:
<!--  Thanks for sending a pull request!  Before submitting:

1. Read our CONTRIBUTING.md guide
2. Rebase your PR if it gets out of sync with main
-->

#### What this PR does

#### Which issue(s) this PR fixes or relates to

Fixes #2772 

#### Checklist

- [x] Tests updated
- [x] Documentation added
- [x] `CHANGELOG.md` updated - the order of entries should be `[CHANGE]`, `[FEATURE]`, `[ENHANCEMENT]`, `[BUGFIX]`

CLAassistant commented:
[![CLA assistant check](https://cla-assistant.io/pull/badge/not_signed)](https://cla-assistant.io/grafana/mimir?pullRequest=3541) <br/>Thank you for your submission! We really appreciate it. Like many open source projects, we ask that you sign our [Contributor License Agreement](https://cla-assistant.io/grafana/mimir?pullRequest=3541) before we can accept your contribution.<br/><sub>You have signed the CLA already but the status is still pending? Let us [recheck](https://cla-assistant.io/check/grafana/mimir?pullRequest=3541) it.</sub>

osg-grafana commented on docs/sources/operators-guide/architecture/components/alertmanager.md:
```diff
@@ -58,9 +58,9 @@ After the tenant uploads an Alertmanager configuration, the tenant can access th
-Specify a fallback configuration using the `-alertmanager.configs.fallback` command-line flag.
+When a tenant doesn't have a Alertmanager configuration, the Grafana Mimir Alertmanager uses a fallback configuration.
+By default, there is always a fallback configuration set.
+The default fallback configuration can be overwritten using the `-alertmanager.configs.fallback` command-line flag.
```
```suggestion
You can overwrite the default fallback configuration via the `-alertmanager.configs.fallback` command-line flag.
```

osg-grafana replied:
Rationale: passive to active voice

JohnnyQQQQ commented on CHANGELOG.md:
```diff
@@ -9,6 +9,7 @@
 * [CHANGE] Distributor: Wrap errors from pushing to ingesters with useful context, for example clarifying timeouts. #3307
 * [CHANGE] The default value of `-server.http-write-timeout` has changed from 30s to 2m. #3346
 * [CHANGE] Reduce period of health checks in connection pools for querier->store-gateway, ruler->ruler, and alertmanager->alertmanager clients to 10s. This reduces the time to fail a gRPC call when the remote stops responding. #3168
+* [CHANGE] Alertmanager: added default configuration so that `-alertmanager.configs.fallback` is not explicitly needed anymore if no config is provided for a tenant. #2772
```
Changed 👍

replay commented:
The CHANGELOG has just been cut to prepare for the next Mimir release. Please rebase `main` and eventually move the CHANGELOG entry added / updated in this PR to the top of the CHANGELOG document. Thanks!

stevesg commented on CHANGELOG.md:
```diff
@@ -7,6 +7,7 @@
 * [CHANGE] Querier: Introduce `-querier.max-partial-query-length` to limit the time range for partial queries at the querier level and deprecate `-store.max-query-length`. #3825
 * [CHANGE] Store-gateway: Remove experimental `-blocks-storage.bucket-store.max-concurrent-reject-over-limit` flag. #3706
 * [CHANGE] Ingester: If shipping is enabled block retention will now be relative to the upload time to cloud storage. If shipping is disabled block retention will be relative to the creation time of the block instead of the mintime of the last block created. #3816
+* [CHANGE] Alertmanager: added default configuration so that `-alertmanager.configs.fallback` is not explicitly needed anymore if no config is provided for a tenant. #2772
```
[nit] This didn't read cleanly to me, tried to make it a bit clearer

```suggestion
* [CHANGE] Alertmanager: added default configuration for `-alertmanager.configs.fallback`. Allows tenants to send alerts without first uploading an Alertmanager configuration. #2772
```

stevesg commented on pkg/alertmanager/multitenant.go:
```diff
@@ -864,7 +891,7 @@ func (am *MultitenantAlertmanager) alertmanagerFromFallbackConfig(ctx context.Co
 	level.Warn(am.logger).Log("msg", "no configuration exists for user; uploading fallback configuration", "user", userID)
 
-	// Upload an empty config so that the Alertmanager is no de-activated in the next poll
+	// Upload an empty config so that the Alertmanager is no de-activated in the next poll.
```
[nit] This was probably just leftover from the last change

```suggestion
	// Upload an empty config so that the Alertmanager is no de-activated in the next poll
```

stevesg reviewed (commented):
Change looks good conceptually. I spoke to @JohnnyQQQQ privately and said that we should move the change for whether or not to upload the default configuration into a separate PR, as I think it has some implications which warrant discussion.

I was thinking that we should add a functional test (either in `multitenant_test.go` or as an integraiton test), to the scenario where we do not set a fallback configuration file, but still attempt to use an alertmanager before it is configured. But I noticed this test already exists [here](https://github.com/grafana/mimir/blob/4141848b47c5c599c8a07d1477f0e882415aab0d/pkg/alertmanager/multitenant_test.go#L945-L957), which in theory should now fail, but doesn't seem to?

The one thing I'm missing

JohnnyQQQQ commented:
There is also one in the [integration folder](https://github.com/grafana/mimir/blob/main/integration/alertmanager_test.go#L188-L190), which is not failing. I'm investigating it.

JohnnyQQQQ commented:
The thing was that the mock that we are using in the test was built manually and not using the constructor, thus no fallback config was computed. I fixed that [here](https://github.com/grafana/mimir/pull/3541/files#diff-5252308a4554ebe5af672884ce76a033819a51531c357f3e9c0c36eda517be1cR105-R109), and now everything works as expected. The tests were failing and I rewrote them to reflect those changes and check for the default config instead of a 412 after a deletion. For the frontend tests the expected behavior is a redirect instead of a 412 after deletion, as the fallback config is used.

stevesg reviewed (approved):
LGTM, I think this is fine to merge once @grafana/docs-metrics approves the docs changes.

osg-grafana commented on pkg/alertmanager/multitenant_test.go:
```diff
@@ -2163,6 +2167,25 @@ receivers:
+	_, err = amconfig.Load(string(fallbackConfig))
+	require.NoError(t, err)
+
+	// If a fallback config file is set, it should return it's content.
```
```suggestion
	// If a fallback configuration file is set, it returns its content.
```

osg-grafana replied:
Spell out terms through for content in plain text. Feel free to ignore this nit if you are pressed for time.

osg-grafana commented on pkg/alertmanager/multitenant_test.go:
```diff
@@ -2163,6 +2167,25 @@ receivers:
 }
 
+func TestMultitenantAlertmanager_computeFallbackConfig(t *testing.T) {
+	// If no fallback config is set, it should return a valid empty config.
```
```suggestion
	// If no fallback configration is set, it returns a valid empty configuration.
```

osg-grafana replied:
To remove ambiguity and raise the tone of confidence, reword instances of "should" throughout to what will or will not happen.

pracucci commented on CHANGELOG.md:
```diff
@@ -8,6 +8,7 @@
 * [CHANGE] Store-gateway: Remove experimental `-blocks-storage.bucket-store.max-concurrent-reject-over-limit` flag. #3706
 * [CHANGE] Ingester: If shipping is enabled block retention will now be relative to the upload time to cloud storage. If shipping is disabled block retention will be relative to the creation time of the block instead of the mintime of the last block created. #3816
 * [CHANGE] Query-frontend: Deprecated CLI flag `-query-frontend.align-querier-with-step` has been removed. #3982
+* [CHANGE] Alertmanager: added default configuration for `-alertmanager.configs.fallback`. Allows tenants to send alerts without first uploading an Alertmanager configuration. #2772
```
We're used to link the PR number. Can you open a follow up PR to change `#2772` to `#3541`, please?

pracucci commented on pkg/alertmanager/multitenant.go:
```diff
@@ -308,6 +302,39 @@ func NewMultitenantAlertmanager(cfg *MultitenantAlertmanagerConfig, store alerts
 
+// ComputeFallbackConfig will load, vaildate and return the provided fallbackConfigFile
+// or return an valid empty default configuration if none is provided.
+func ComputeFallbackConfig(fallbackConfigFile string) ([]byte, error) {
```
[nit] Not needed to export it. Could be `computeFallbackConfig()`.

pracucci commented:
Nice job, thanks!
